pandas/io/parsers.py

```diff
@@ -1146,18 +1146,27 @@ def __init__(self, src, **kwds):
             else:
                 self.names = lrange(self._reader.table_width)
 
-        # If the names were inferred (not passed by user) and usedcols is
-        # defined, then ensure names refers to the used columns, not the
-        # document's columns.
-        if self.usecols and passed_names:
-            col_indices = []
-            for u in self.usecols:
-                if isinstance(u, string_types):
-                    col_indices.append(self.names.index(u))
-                else:
-                    col_indices.append(u)
-            self.names = [n for i, n in enumerate(self.names)
-                          if i in col_indices]
+        # gh-9755
+        #
+        # need to set orig_names here first
+        # so that proper indexing can be done
+        # with _set_noconvert_columns
+        #
+        # once names has been filtered, we will
+        # then set orig_names again to names
+        self.orig_names = self.names[:]
+
+        if self.usecols:
+            if len(self.names) > len(self.usecols):
+                col_indices = []
+                for u in self.usecols:
+                    if isinstance(u, string_types):
```

Comment:
that will should fail, how are you supposed to interpret a mixed ``usecols``? what if you columns are: ``['3', '4', 'b', '1', '2']``.

If don't think we have restrictions on ``usecols`` ATM, but they should either be all positional (e.g. integer dtype), or not.

In fact, pls you put an ambiguous ``usecols`` in your last PR: https://github.com/pydata/pandas/pull/12551.

pls fix that. So we need validation on ``usecols``. 

you can do something like:

```
In [8]: pd.lib.infer_dtype([1,2,3])
Out[8]: 'integer'

In [9]: pd.lib.infer_dtype(['1','2','3'])
Out[9]: 'string'

In [10]: pd.lib.infer_dtype(['1','2',3])
Out[10]: 'mixed-integer'
```

and raise on [10])